### PR TITLE
[FLINK-12955] HBase LookupableTableSource

### PR DESCRIPTION
**Pull request by chancelq**
## What is the purpose of the change

Add HBase LookupableTableSource.

## Brief change log 
add a LookupableTableSource implementation in HBase Connector.

## Verifying this change
UT and ITCase cover the basic logic.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseStreamTableSource.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import java.io.IOException;
+
+/**
+ * A HBase TableSource for stream table.
+ */
+public class HBaseStreamTableSource implements StreamTableSource<Row>, LookupableTableSource<Row> {
+
+	private final HBaseTableContext hBaseTableContext;
+
+	public HBaseStreamTableSource(HBaseTableContext hBaseTableContext) {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseStreamTableSource.java]: Can be package-private

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseStreamTableSource.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		throw new UnsupportedOperationException("HBase table can not convert to DataStream currently.");
+	}
+
+	@Override
+	public DataType getProducedDataType() {
+		return hBaseTableContext.getTableSchema().toRowDataType();
+	}
+
+	@VisibleForTesting
+	public HBaseTableContext gethBaseTableContext() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseStreamTableSource.java]: Can be package-private. Rename to `getHBaseTableContext`

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			if (this.rowkey.equals(columnNames[idx])) {
+				rowKeySourceIndex = idx;
+				break;
+			}
+		}
+		Preconditions.checkArgument(rowKeySourceIndex > -1, "rowKeySourceIndex must > -1, no rowkey?");
+		return rowKeySourceIndex;
+	}
+
+	public String gethTableName() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java]: Can be package-private. Rename to `getHTableName`

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+		return hTableName;
+	}
+
+	public TableSchema getTableSchema() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java]: Can be package-private.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+		return this.tableSchema;
+	}
+
+	public Configuration getHbaseClientConf() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java]: Can be package-private

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+		return hbaseClientConf;
+	}
+
+	public HBaseTableSchema gethBaseTableSchema() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java]: Can be package-private

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+		return hBaseTableSchema;
+	}
+
+	public String getRowkey() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java]: Can be package-private. Rename to `getRowKey`

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+		return rowkey;
+	}
+
+	public TypeInformation<?> getRowKeyType() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java]: Can be package-private

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+		return this.rowKeyType;
+	}
+
+	public int getRowKeyIndex() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java]: Can be package-private

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+		return this.rowkeyIndex;
+	}
+
+	public List<Integer> getQualifierIndexes() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableContext.java]: Can be package-private

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableFactory.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+		// and append or overwrite configuration using user params from client-side which has highest priority
+		if (null != userParams) {
+			for (Map.Entry<String, String> entry : userParams.entrySet()) {
+				clientConfiguration.set(entry.getKey(), entry.getValue());
+			}
+		}
+		return clientConfiguration;
+	}
+
+	String hbaseVersion() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseTableFactory.java]: Can be private method.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseValidator.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+public class HBaseValidator extends ConnectorDescriptorValidator {
+	public static final String COLUMNFAMILY_QUALIFIER_DELIMITER_PATTERN = ":";
+
+	public static final String CONNECTOR_TYPE_VALUE_HBASE = "hbase";
+	public static final String CONNECTOR_VERSION_VALUE_143 = "1.4.3";
+
+	public static final String CONNECTOR_HBASE_CLIENT_PARAM_PREFIX = "hbase.*";
+	public static final String CONNECTOR_HBASE_TABLE_NAME = "connector.table-name";
+	public static final String CONNECTOR_QUALIFIER_DELIMITER = "connector.qualifier-delimiter";
+	public static final String CONNECTOR_ROW_KEY = "connector.row-key";
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/table/HBaseValidator.java]: All above final fields in this class could be package-private.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseBytesSerializer.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+import java.io.Serializable;
+import java.io.UnsupportedEncodingException;
+
+/**
+ * HBaseBytesSerializer gives support to serialize into or deserialize from HBase bytes data.
+ * Currently only support simple java data types, include:
+ * byte[], String, Byte, Short, Integer, Long, Float, Double, Boolean, Timestamp, Date, Time, BigDecimal
+ */
+public class HBaseBytesSerializer implements Serializable {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseBytesSerializer.java]: It seems this class is never used. What's the purpose of introducing it?

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+import java.sql.Time;
+import java.sql.Timestamp;
+
+/**
+ * A utility class to process data exchange with HBase type system.
+ */
+@Internal
+public class HBaseTypeUtils {
+
+	public static final String DEFAULT_CHARSET = "UTF-8";
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java]: Can be package-private.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+	private static final byte[] EMPTY_BYTES = new byte[]{};
+
+	//convert to external Java type from HBase bytes
+	public static Object deserializeToObject(
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java]: Can be package-private.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+				return Bytes.toBytes((BigDecimal) value);
+			case 13:
+				return ((BigInteger) value).toByteArray();
+
+			default:
+				throw new IllegalArgumentException("unsupported type index:" + typeIdx);
+		}
+	}
+
+	public static int getTypeIndex(Class<?> clazz) {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java]: Can be private.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+			case 13:
+				return new BigInteger(value);
+
+			default:
+				throw new IllegalArgumentException("unsupported type index:" + typeIdx);
+		}
+	}
+
+	//convert to HBase bytes from the external type
+	public static byte[] serializeFromObject(
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java]: Can be package-private.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+		} else if (BigDecimal.class.equals(clazz)) {
+			return 12;
+		} else if (BigInteger.class.equals(clazz)) {
+			return 13;
+		} else {
+			throw new IllegalArgumentException("unsupported type class:" + clazz);
+		}
+	}
+
+	public static int getTypeIndex(TypeInformation typeInfo) {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java]: Can be package-private.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+		return getTypeIndex(typeInfo.getTypeClass());
+	}
+
+	public static boolean isSupportedType(Class<?> clazz) {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/HBaseTypeUtils.java]: Never used, could be safely deleted.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/WritableSerializer.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+	/**
+	 * Serialize writable byte[].
+	 *
+	 * @param <T>      the type parameter
+	 * @param writable the writable
+	 * @return the byte [ ]
+	 * @throws IOException the io exception
+	 */
+	public static <T extends Writable> byte[] serializeWritable(T writable) throws IOException {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/WritableSerializer.java]: Can be package-private.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/WritableSerializer.java [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+
+	/**
+	 * Deserialize writable.
+	 *
+	 * @param <T>      the type parameter
+	 * @param writable the writable
+	 * @param bytes    the bytes
+	 * @throws IOException the io exception
+	 */
+	public static <T extends Writable> void deserializeWritable(T writable, byte[] bytes)
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/connectors/hbase/util/WritableSerializer.java]: Can be package-private.

**Comment by zentol** [comment]
CI report for commit ae1363bb0188a922190234d5f07b02fac1a175e7: FAILURE [Build](https://travis-ci.org/flink-ci/flink-ci/builds/554872254?utm_source=github_status&utm_medium=notification)

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseLookupFunction.java [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+	private static final long serialVersionUID = 1L;
+
+	private final String hTableName;
+	private byte[] serializedConfig;
+	private transient Connection hConnection;
+	private transient HTable table;
+	private final RowParser<Result> rowParser;
+	private final TypeInformation<Row> resultType;
+
+	public HBaseLookupFunction(
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseLookupFunction.java]: Change the visibility to package-private

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseLookupFunction.java [hunk 23]
```diff
@@ -0,0 +1,153 @@
...
+		TableSchema tableSchema,
+		RowParser<Result> rowParser) throws IOException {
+		this.resultType = tableSchema.toRowType();
+		this.rowParser = rowParser;
+		this.hTableName = hTableName;
+		// Configuration is not serializable
+		this.serializedConfig = HBaseConfigurationUtil.serializeConfiguration(configuration);
+	}
+
+	public void eval(Object rowKey) throws IOException {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseLookupFunction.java]: This method is never used. Remove it.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java [hunk 24]
```diff
@@ -179,8 +185,57 @@ void setCharset(String charset) {
...
+	public String getStringCharset() {
 		return this.charset;
 	}
 
+	/**
+	 * A helper method returns a flat qualifiers list in defining order.
+	 * Qualifier info is a Tuple3 representation:
+	 * (byte[] columnFamily, byte[] qualifier, TypeInformation typeInfo)
+	 */
+	public List<Tuple3<byte[], byte[], TypeInformation<?>>> getFlatByteQualifiers() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java]: Change the method visibility to package-private

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java [hunk 25]
```diff
@@ -179,8 +185,57 @@ void setCharset(String charset) {
...
+	}
+
+	/**
+	 * Adds a column defined by family, qualifier, and type to the table schema.
+	 *
+	 * @param family    the family name
+	 * @param qualifier the qualifier name
+	 * @param type the data type info of the qualifier
+	 */
+	public void addColumn(String family, String qualifier, TypeInformation<?> type) {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSchema.java]: Change the method visibility to package-private

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSource.java [hunk 26]
```diff
@@ -75,7 +83,7 @@ public HBaseTableSource(Configuration conf, String tableName) {
 		this.hBaseSchema = new HBaseTableSchema();
 	}
 
-	private HBaseTableSource(Configuration conf, String tableName, TableSchema tableSchema) {
+	public HBaseTableSource(Configuration conf, String tableName, TableSchema tableSchema) {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSource.java]: Change the method visibility to package-private (no need to be public)

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseValidator.java [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * The validator for HBase.
+ * More features to be supported, e.g., batch read/write, async api(support from hbase version 2.0.0), Caching for LookupFunction.
+ */
+@Internal
+public class HBaseValidator extends ConnectorDescriptorValidator {
+	public static final String COLUMNFAMILY_QUALIFIER_DELIMITER_PATTERN = ":";
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseValidator.java]: This field is never used, other fields don't need to be public

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/parser/NestedRowParser.java [hunk 28]
```diff
@@ -0,0 +1,186 @@
...
+			for (byte[] qualifier : qualifiers[f]) {
+				get.addColumn(family, qualifier);
+			}
+		}
+		return get;
+	}
+
+	@Override
+	public Row parseToRow(Result result, Object rowKey) throws IOException {
+		Row row = new Row(fieldLength);
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/parser/NestedRowParser.java]: Never used local field, remove it.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/parser/RowParser.java [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+	/**
+	 * init a RowParser.
+	 */
+	default void init() {}
+
+	/**
+	 * create Get from input parameter.
+	 * NOTICE: now only support key.
+	 */
+	Get createGet(Object in) throws IOException;
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/parser/RowParser.java]: This interface is implemented but seems never used.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/parser/RowParser.java [hunk 30]
```diff
@@ -0,0 +1,47 @@
...
+
+	/**
+	 * parse HBase result to a Row.
+	 */
+	Row parseToRow(T result, Object rowKey) throws IOException;
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/parser/RowParser.java]: This interface is implemented but seems never used.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/util/HBaseBytesSerializer.java [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+
+import java.io.Serializable;
+import java.io.UnsupportedEncodingException;
+
+/**
+ * HBaseBytesSerializer gives support to serialize into or deserialize from HBase bytes data.
+ * Currently only support simple java data types, include:
+ * byte[], String, Byte, Short, Integer, Long, Float, Double, Boolean, Timestamp, Date, Time, BigDecimal
+ */
+public class HBaseBytesSerializer implements Serializable {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/util/HBaseBytesSerializer.java]: This class is never used, consider to remove it.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/util/HBaseTypeUtils.java [hunk 32]
```diff
@@ -0,0 +1,165 @@
...
+		} else {
+			throw new IllegalArgumentException("unsupported type class:" + clazz);
+		}
+	}
+
+	public static int getTypeIndex(TypeInformation typeInfo) {
+		return getTypeIndex(typeInfo.getTypeClass());
+	}
+
+	public static boolean isSupportedType(Class<?> clazz) {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/util/HBaseTypeUtils.java]: The method seems to be never used, consider to remove it.

File: flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/util/WritableSerializer.java [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+
+	/**
+	 * Serialize writable byte[].
+	 *
+	 * @param <T>      the type parameter
+	 * @param writable the writable
+	 * @return the byte [ ]
+	 * @throws IOException the io exception
+	 */
+	public static <T extends Writable> byte[] serializeWritable(T writable) throws IOException {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/util/WritableSerializer.java]: Access can be package-private, no need to be public.

File: flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/HBaseLookupFunctionITCase.java [hunk 34]
```diff
@@ -0,0 +1,247 @@
...
+	}
+
+	/**
+	 * Allows the tests to use {@link ExecutionEnvironment#getExecutionEnvironment()} but with a
+	 * configuration that limits the maximum memory used for network buffers since the current
+	 * defaults are too high for Travis-CI.
+	 */
+	private abstract static class LimitNetworkBuffersTestEnvironment extends ExecutionEnvironment {
+
+		public static void setAsContext() {
```
Comment (carp84) [inline flink-connectors/flink-hbase/src/test/java/org/apache/flink/addons/hbase/HBaseLookupFunctionITCase.java]: Never used, consider removing it.

**Comment by wuchong** [comment]
Thanks @chancelq for the great effort. With discussed with @chancelq , we will continue the work in #9045 .

**Pull request closed without merging**